src/kvaser_can_bridge.cpp

```diff
@@ -19,89 +19,74 @@ using namespace AS::CAN;
 int bit_rate = 500000;
 int hardware_id = 0;
 int circuit_id = 0;
-bool global_keep_going = true;
-std::mutex keep_going_mut;
 KvaserCan can_reader, can_writer;
 ros::Publisher can_tx_pub;
 
 void can_read()
 {
-  uint32_t id;
-  uint8_t msg[8] = {};  // zero-initialize array
-  uint32_t size;
-  bool extended;
-  uint64_t t;
-
-  const std::chrono::milliseconds loop_pause = std::chrono::milliseconds(10);
-  bool keep_going = true;
-
-  // Set local to global value before looping.
-  keep_going_mut.lock();
-  keep_going = global_keep_going;
-  keep_going_mut.unlock();
-
   ReturnStatuses ret;
 
-  while (keep_going)
+  while (true)
   {
-    std::chrono::system_clock::time_point next_time = std::chrono::system_clock::now();
-    next_time += loop_pause;
-
     if (!can_reader.isOpen())
     {
       ret = can_reader.open(hardware_id, circuit_id, bit_rate, false);
 
       if (ret != ReturnStatuses::OK)
+      {
         ROS_ERROR_THROTTLE(0.5, "Kvaser CAN Interface - Error opening reader: %d - %s",
           static_cast<int>(ret), KvaserCanUtils::returnStatusDesc(ret).c_str());
+        break;
+      }
     }
-    else
+
+    if (can_reader.isOpen())
     {
-      while (true)
-      {
-        ret = can_reader.read(&id, msg, &size, &extended, &t);
+      CanMsg msg;
+
+      ret = can_reader.read(&msg);
 
-        if (ret  == ReturnStatuses::OK)
+      if (ret  == ReturnStatuses::OK)
+      {
+        // Only publish if msg is not CAN FD.
+        // Also, only publish if msg contains
+        // data, is RTR, or is error frame.
+        if (!msg.flags.fd_msg &&
```

Comment:
I think can messages without any data in them are considered valid (for instance just the reception of a message may trigger an event/action of some sort).  I haven't seen any of our current sensors do this, but should we keep this more open?  Is there some reason for only allowing non-empty messages through?

Reply:
Are we intending to support CAN FD? If we are then we probably shouldn't be filtering out FD messages here. If not then we can probably skip some of the changes in kvaser_interface.

Reply:
My intent is to have the kvaser_interface support CAN FD but, currently, can_msgs does not support CAN FD (the data member of the Frame msg is not large enough and changing this would require evaluating all of the changes to ros_canopen). So, we have a few options:

1. Add CAN FD Frame message to can_msgs and publish them on a separate topic from regular can Frames.
2. Modify the CAN Frame message and evaluate/fix any changes that will occur to ros_canopen.
3. Create our own CAN Frame message which supports FD frames and update all of our drivers to use that instead.
4. Support FD with the interface, only support traditional frames with the ROS layer, and fix the problem in the future.

Reply:
My first-impression opinion is that option 2 is probably the way to go. Might as well abstract it for everything downstream of this at once.